### ISSUE TITLE
cmd/compile: avoid memclr for provable overwritten slices

### ISSUE THREAD

**Issue by dsnet** [golang/go#57153]
Consider [the following pattern](https://go.dev/play/p/EqqpHq6_Rzt):
```go
var n int
for _, src := range source {
	n += len(src)
}
b := make([]byte, 0, n)
for _, src := range source {
	b = append(b, src...)
}
sink = b
```
which often occurs when merging segmented buffers.

Running this benchmark shows that the CPU time is dominated by `runtime.memclrNoHeapPointers`, which occupies 44% of the time, while `runtime.memmove` occupies 36% of the time.

The compiler should be able to prove that every byte of this slice is overwritten, and thus avoid the call to `memclr` entirely.

If detecting this pattern is too difficult, we could instead special-case `bytes.Join` such that the [internal call to `make`](https://github.com/golang/go/blob/9431237d772f4a3b670f8533868d6df0488688a9/src/bytes/bytes.go#L536) avoids the `memclr`. Thus, the pattern above could be replaced with `bytes.Join(source, nil)`.

\cc @randall77 @martisch @lavalamp

**Comment by go101** [comment]
It is hard to implement this for all cases. But it might be a good idea to do some special optimizations for the later `slices` std package.

**Comment by lavalamp** [comment]
Re:

```
var b []byte
for _, src := range source {
	b = append(b, src...)
}
```

Requiring that to be the canonical pattern has the weird property that someone pre-allocating with the "right" size will get worse performance.

When I encountered this my first wish was for append to solve the problem via a double definition:

```
func append[T any](base []T, args... T)
func append[T any](base []T, args... []T)
```

(I'm not claiming that's a great solution, just a data point about what one go-user would have found intuitive)

**Comment by lavalamp** [comment]
It seems like an alternative approach might be "lazy clearing" -- track an index i such that everything prior to that index doesn't need clearing and everything after it is uninitialized.

After an allocation just avoid clearing the remaining part of the array until one of these conditions is met:
* A read from the uncleared part of the array
* A write that doesn't start at or before i (and in this case i is incremented to become the end of the written data)
* The end of the function
* A reference to the array potentially escapes

**Comment by dsnet** [comment]
> Generally I would not want to open the Pandora's box of giving std lib functions like bytes.Join special runtime acces

Heh... that's exactly the approach I took in https://go.dev/cl/456336. I personally think it's fairly palatable since `bytes` and `strings` already depend on unsafe operations (especially through the `internal/bytealg`) package.

As part of that CL, I also optimized `strings.Builder`, which will have many downstream benefits. The way `strings.Builder` uses the `[]byte` would be very difficult for the compiler to prove that the unused capacity never leaks to the end user.

In general, this provides 2x performance gains for sufficiently large buffers. This makes sense since we're removing an entire write pass through the memory.

**Comment by go101** [comment]
```Go
func append[T any](base []T, args... []T)
```

has ambiguity if the type argument for `T` is an interface type.

**Comment by dsnet** [comment]
> good idea to do some special optimizations for the later slices std package.

I don't think this belongs in the `slices` package since this optimization only works for element types that do not contain pointers. There would need to be non-trivial runtime magic in the slices package to make this work (even more so than the `internal/bytealg` approach I took in https://go.dev/cl/456336).

**Comment by prattmic** [comment]
cc @golang/compiler

**Comment by dsnet** [comment]
@martisch: In principle, I agree with you. However, I'm not sure compiler optimization is going to fix all cases. For example, it seems like a hard (though solvable) problem statically detecting that the `make([]byte, ...)` in `strings.Builder.Grow` never leaks uninitialized bytes to the end user. I fear that we're allowing the perfect become the enemy of the good.

**Comment by andig** [comment]
I may misunderstand this. @martisch seems to imply that users will use "linknaming into runtime" once the bytealg function is available. After >5 years of go I'm aware of unsafe (though I've never needed to use it) but I have only a faint idea of linknaming. I seem to recall a Go issue where the go team made clear that using it is "your problem". I may of cause be wrong, but isn't bytealg with potential misuse a much smaller foot gun (given it's not promoted as Go api)  than an unsafe function that actually invites you to do so?
